picoscope/ps6000a.py

```diff
@@ -0,0 +1,963 @@
+        numSamplesReturned = c_uint64()
+        numSamplesReturned.value = numSamples
+        overflow = c_int16()
+        m = self.lib.ps6000aGetValues(
+            c_int16(self.handle),
+            c_uint64(startIndex),
+            byref(numSamplesReturned),
+            c_uint64(downSampleRatio),
+            c_enum(downSampleMode),
+            c_uint64(segmentIndex),
+            byref(overflow))
+        self.checkResult(m)
+        return (numSamplesReturned.value, overflow.value)
+
+    # Misc
+    def _lowLevelGetTriggerTimeOffset(self, segmentIndex):
+        time = c_int64()
+        timeUnits = c_enum()
+
+        m = self.lib.ps6000aGetTriggerTimeOffset64(
+            c_int16(self.handle),
+            byref(time),
+            byref(timeUnits),
+            c_uint64(segmentIndex))
+        self.checkResult(m)
+
+        try:
+            return time.value * self.TIME_UNITS[timeUnits.value]
+        except KeyError:
+            raise TypeError("Unknown timeUnits %d" % timeUnits.value)
+
+    "Updates"
+
+    def _lowLevelCheckForUpdate(self):
+        """
+        Check whether a firmware update for the device is available.
+
+        Returns
+        -------
+        firmwareInfos : None
+            not implemented: A struct with information.
+        number : int
+            Number of elements in the structure.
+        required : bool
+            Whether an update is required or not.
+        """
+        # TODO what to do with the struct?
+        number = c_int16()
+        required = c_uint16()
+        m = self.lib.ps6000aCheckForUpdate(c_int16(self.handle), c_void_p(),
+                                           byref(number), byref(required))
+        self.checkResult(m)
+        return None, number, required
+
+    def _lowLevelStartFirmwareUpdate(self, function):
+        callback = updateFirmwareProgress(function)
+        m = self.lib.ps6000aStartFirmwareUpdate(c_int16(self.handle), callback)
+        self.checkResult(m)
+
+    "Misc"
+
+    ##################################
+    # TODO verify functions below here in the manual
+    # TODO ensure all relevant functions are in here
+    ##################################
+
+    "Measurement"
+
+    # Complicated triggering
+    # need to understand structs for some of this to work
+    def _lowLevelGetValuesTriggerTimeOffsetBulk():
+        pass
+
+    def _lowLevelSetTriggerChannelConditions():
+        pass
+
+    def _lowLevelSetTriggerChannelDirections():
+        pass
+
+    def _lowLevelSetTriggerChannelProperties():
+        pass
+
+    def _lowLevelSetTriggerDelay():
+        pass
+
+    def _lowLevelSetTriggerDigitalPortProperties(self):
+        pass
+
+    # Optional input triggering: PulseWidthQualifier
+    def _lowLevelSetPulseWidthQualifierConditions(self):
+        pass
+
+    def _lowLevelSetPulseWidthQualifierDirections(self):
+        pass
+
+    def _lowLevelSetPulseWidthQualifierProperties(self):
+        pass
```

Comment:
It will be best, when we are done, to simply remove these functions so they cause a visible to the end user as opposed to silently failing.